crates/matrix-sdk-ui/src/timeline/builder.rs

```diff
@@ -502,3 +313,229 @@ impl TimelineBuilder {
         Ok(timeline)
     }
 }
+
+/// The task that handles the pinned event IDs updates.
+async fn pinned_events_task<S>(pinned_event_ids_stream: S, inner: TimelineController)
+where
+    S: Stream<Item = Vec<OwnedEventId>>,
+{
+    pin_mut!(pinned_event_ids_stream);
+
+    while pinned_event_ids_stream.next().await.is_some() {
+        if let Ok(events) = inner.reload_pinned_events().await {
+            inner
+                .replace_with_initial_remote_events(
+                    events.into_iter(),
+                    RemoteEventOrigin::Pagination,
+                )
+                .await;
+        }
+    }
+}
+
+/// The task that handles the [`RoomEventCacheUpdate`]s.
+async fn room_event_cache_updates_task(
+    room_event_cache: RoomEventCache,
+    timeline_controller: TimelineController,
+    mut event_subscriber: RoomEventCacheListener,
+    is_live: bool,
+) {
+    trace!("Spawned the event subscriber task.");
+
+    loop {
+        trace!("Waiting for an event.");
+
+        let update = match event_subscriber.recv().await {
+            Ok(up) => up,
+            Err(RecvError::Closed) => break,
+            Err(RecvError::Lagged(num_skipped)) => {
+                warn!(num_skipped, "Lagged behind event cache updates, resetting timeline");
+
+                // The updates might have lagged, but the room event cache might have
+                // events, so retrieve them and add them back again to the timeline,
+                // after clearing it.
+                let (initial_events, _stream) = room_event_cache.subscribe().await;
+
+                timeline_controller
+                    .replace_with_initial_remote_events(
+                        initial_events.into_iter(),
+                        RemoteEventOrigin::Sync,
```

Comment:
It becomes `::Cache` in https://github.com/matrix-org/matrix-rust-sdk/pull/4795. It should be updated before merge.